### ISSUE TITLE
T3-T4 retrieval: section/work-level metadata routing for RUTH + OES

### ISSUE THREAD

**Issue by krisztiankoos** [learn-ukrainian/learn-ukrainian.github.io#1341]
# T3-T4 retrieval: section/work-level metadata routing for RUTH + OES

Part of EPIC #1335. Depends on #1337 (schema). LOWER PRIORITY — RUTH (115 modules) and OES (102 modules) are months out behind A1/A2/B1.

## Scope

Build the COMPILE-layer retrieval path for archaic Ukrainian (T3 = Middle Ukrainian / RUTH track, T4 = Old East Slavic / OES track). Per discussion consensus: NO dense embeddings, NO hybrid — both empirically failed in `#1101`. Pure metadata-routed section/work-level retrieval.

## Why a separate pipeline

`#1101` benchmark numbers force the split:

| Tier | Dense Recall@10 | Hybrid Recall@10 |
|---|---|---|
| Middle Ukrainian | 0.600 | 0.467 |
| Old East Slavic | 0.400 | 0.300 |

Modern embedding models (BGE-M3, e5, jina-v3) are trained predominantly on contemporary web text. Archaic Slavic orthography, vocabulary, and morphology are out-of-distribution. The available historical-Slavic models (e.g. `BERTislav`, https://huggingface.co/npedrazzini/BERTislav) are fill-mask baselines, NOT validated retrieval models. Until an in-domain historical retrieval benchmark exists, dense-on-archaic is an experiment, not architecture.

The retrieval unit also changes. Modern textbooks have clear section/chapter structure; archaic chronicles and Middle Ukrainian texts have manuscript divisions, tales-within-chronicles, and editorial sections. The metadata-first router is more important than the search method.

### AC1 — `literary_sections` table

Add to `data/sources.db`:

```sql
CREATE TABLE literary_sections (
    section_id      INTEGER PRIMARY KEY,
    work_id         TEXT NOT NULL,        -- e.g. "povest_vremennykh_let"
    work_title      TEXT NOT NULL,        -- "Повість временних літ"
    period          TEXT NOT NULL,        -- "old_east_slavic" | "middle_ukrainian" | "modern_literary"
    genre           TEXT NOT NULL,        -- "chronicle" | "religious" | "legal" | "poetry" | "prose" | "drama"
    section_title   TEXT,                 -- e.g. "Лѣто 6360" or "Слово о князи Ігори"
    section_number  TEXT,
    chunk_count     INTEGER NOT NULL,
    full_text       TEXT NOT NULL,
    UNIQUE (work_id, section_title)
);
CREATE INDEX idx_litsec_period_genre ON literary_sections (period, genre);
CREATE INDEX idx_litsec_work ON literary_sections (work_id);
```

Add to existing `literary_texts` table:
```sql
ALTER TABLE literary_texts ADD COLUMN parent_section_id INTEGER REFERENCES literary_sections(section_id);
```

### AC2 — Section extraction

`scripts/wiki/extract_literary_sections.py`:
- Read existing `literary_texts` chunks
- Group by `(work_id, section_title)` — section_title may be embedded in `title` column with prefixes like "Лѣто" (year-marked chronicle entries)
- Extract `period` from existing source metadata (literary collection ingest already tags this — verify via `scripts/wiki/build_sources_db.py`)
- Extract `genre` from source metadata or by manual mapping table for the ~30-50 distinct works in the literary corpus
- Idempotent (`--rebuild` drops and re-creates)

### AC3 — `search_sources(strategy="archaic_metadata")` implementation

`scripts/wiki/sources_db.py`:
- Strategy parameter:
  - `track="ruth"` → period filter `IN ("middle_ukrainian", "old_east_slavic")` (Middle UK is primary, OES allowed for chronicle prehistory)
  - `track="oes"` → period filter `=="old_east_slavic"` strictly
- Query construction reuses #1337's bucket-A/bucket-B but applies to FTS5 over `literary_texts` (chunk-level), then groups by `parent_section_id`
- Genre routing: callers can pass `--genre chronicle,religious` to scope further
- NO dense re-rank step. Pure FTS5 + metadata. Return top-K sections by FTS5 BM25 score within metadata-filtered candidate set.

### AC4 — Wiki-track wiring

`scripts/wiki/enrichment.py::enrich_sources`:
- For tracks `ruth` and `oes`: call `search_sources(strategy="archaic_metadata")`
- Each returned section's `full_text` becomes one source block (analogous to T1-T2 pattern)
- Char budget enforcement per existing `LEVEL_CHAR_BUDGETS`

### AC5 — End-to-end smoke test

`tests/wiki/test_t3_t4_pipeline.py`:
- Pick one OES wiki slug (TBD — a chronicle-period article from `wiki/periods/`) and one RUTH slug
- Assert returned sections have correct period/genre per the track
- Assert NO dense encoding is invoked (mock the BGE-M3 module and assert it's never called for archaic strategy)

### AC6 — Lint + tests

`.venv/bin/ruff check` clean. `.venv/bin/pytest tests/wiki/test_t3_t4_pipeline.py tests/wiki/test_extract_literary_sections.py -v` green.

## Out of scope

- Reintroducing dense-on-archaic. Explicitly rejected per discussion + #1101.
- Curating the `genre` mapping for all literary works exhaustively — start with the top-10 most-cited works, expand as RUTH/OES wiki articles get built.
- T3-T4 wiki article rebuild — that's track-build work, not retrieval-pipeline work.

## Owner / dispatch

Codex, AFTER #1338 (T1-T2) ships and validates. RUTH/OES tracks are not on the immediate critical path; this ticket can stay in the queue until A1/A2/B1 are well underway.

## Closing criteria

Schema migrated. Extraction passes on full literary corpus. End-to-end smoke test green. One sample RUTH and one sample OES wiki article re-compiles cleanly with new pipeline.

**Comment by krisztiankoos** [comment]
#1343 update: after refactoring the benchmark off Qdrant and onto `data/sources.db`, I do **not** have evidence that a dense embedder should replace the metadata-first T3-T4 strategy.

Current recommendation for the archaic pipeline: keep `#1341` metadata-first, with no dense embedder as the primary retrieval stage.

Numbers from the only full dense run that completed locally (`BAAI/bge-m3`, `sample_size=1000`):
- `middle_ukrainian`: Recall@10 `0.50`, nDCG@10 `0.2617`
- `old_east_slavic`: Recall@10 `0.30`, nDCG@10 `0.2693`

The legacy BGE hybrid baseline did not rescue the archaic tiers:
- `middle_ukrainian` Recall@10 fell to `0.45`
- `old_east_slavic` Recall@10 fell to `0.20`

Blocked / partial contenders from #1343:
- `Qwen/Qwen3-Embedding-0.6B`: blocked in the full run by MPS memory pressure.
- `google/embeddinggemma-300m`: gated Hugging Face repo (`401 Unauthorized`) in this environment.
- `jinaai/jina-embeddings-v3`: smoke run only, not enough to overturn the architecture.

So the handoff from #1343 to #1341 is: keep metadata-only / metadata-first for now, and reopen dense retrieval only when a model completes the same local benchmark and materially beats these archaic numbers.

**Comment by krisztiankoos** [comment]
Closing as **deferred**. Decision 2026-04-19 late afternoon (see session handoff `docs/session-state/current.md`).

**Rationale**: Per the #1345 bakeoff + #1101 archaic-tier numbers, we do not have a retrieval approach for T3 (Middle Ukrainian / RUTH) or T4 (Old East Slavic / OES) that clears the quality bar worth locking into architecture.

- `#1101` dense Recall@10: Middle Ukrainian 0.600 · Old East Slavic 0.400
- `#1101` hybrid Recall@10: Middle Ukrainian 0.467 · Old East Slavic 0.300
- Modern embedders (BGE-M3, e5, jina-v3, EmbeddingGemma) are out-of-distribution on archaic Slavic orthography/morphology. The bakeoff confirmed this: EmbeddingGemma scored 0.03 R@10 on the MODERN tier alone, and no candidate archaic-specialized retrieval model was available in this environment.
- BERTislav and similar historical-Slavic models are fill-mask baselines, not validated retrieval models. Shipping metadata-only as a stopgap would lock in mediocrity and make it harder to justify a proper rebuild later.

**Non-blocking context**: RUTH (115 modules) and OES (102 modules) are months out behind A1/A2/B1. There is no content pipeline on these tracks actively starved by this ticket.

**Revisit**: 2026 Q3, or earlier if a credible candidate emerges:
- A sovereign UK LLM with historical corpus pretraining
- A validated historical-Slavic retrieval benchmark we can A/B against
- An in-domain fine-tune of BGE-M3 or similar on our existing archaic corpus

When revisiting, the schema in this ticket's AC1 (`literary_sections` table) is still the right structure — the metadata-router design holds, only the retrieval method is deferred.

Tracking: EPIC #1335 acknowledges this gap. Will not mark EPIC complete until T3-T4 is either shipped or formally scoped out of the multi-tier design.
